crates/shared/src/http_client.rs

```diff
@@ -19,11 +23,141 @@ pub async fn response_body_with_size_limit(
     Ok(bytes)
 }
 
+#[derive(Debug, Clone)]
+pub struct RateLimitingStrategy {
+    drop_requests_until: Instant,
+    /// How many requests got rate limited in a row.
+    times_rate_limited: u64,
+    back_off_growth_factor: f64,
+    min_back_off: Duration,
+    max_back_off: Duration,
+}
+
+impl RateLimitingStrategy {
+    pub fn try_new(
+        back_off_growth_factor: f64,
+        min_back_off: Duration,
+        max_back_off: Duration,
+    ) -> Result<Self> {
+        ensure!(
+            back_off_growth_factor.is_normal(),
+            "back_off_growth_factor must be a normal f64"
+        );
+        ensure!(
+            back_off_growth_factor > 1.0,
+            "back_off_growth_factor needs to be greater than 1.0"
+        );
+        ensure!(
+            min_back_off <= max_back_off,
+            "min_back_off needs to be <= max_back_off"
+        );
+        Ok(Self {
+            drop_requests_until: Instant::now(),
+            times_rate_limited: 0,
+            back_off_growth_factor,
+            min_back_off,
+            max_back_off,
+        })
+    }
+}
+
+impl RateLimitingStrategy {
+    /// Resets back off and stops rate limiting requests.
+    pub fn response_ok(&mut self) {
+        self.times_rate_limited = 0;
+        self.drop_requests_until = Instant::now();
+    }
+
+    /// Calculates back off based on how often we got rate limited in a row.
+    fn get_current_back_off(&self) -> Duration {
+        let mut back_off = self.min_back_off;
+        for _ in 0..self.times_rate_limited {
+            // do this iteratively to reasonably prevent panics on overflow
+            back_off = back_off.mul_f64(self.back_off_growth_factor);
+            if back_off >= self.max_back_off {
+                return self.max_back_off;
+            }
+        }
+        back_off
+    }
+
+    /// Returns updated back off if no other thread increased it in the mean time.
+    pub fn response_rate_limited(&mut self, previous_rate_limits: u64) -> Option<Duration> {
+        if self.times_rate_limited != previous_rate_limits {
+            // Don't increase back off if somebody else already updated it in the meantime.
+            return None;
```

Comment:
Instead of returning `None`, should we return the current backoff here, even if we don't increment the counter?